alg/teca_model_segmentation.py

```diff
@@ -0,0 +1,227 @@
+import sys
+import teca_py
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+class teca_model_segmentation:
+    """
+    Given an input field of integrated vapor transport,
+    calculates the probability of AR presence at each gridcell.
+    """
+    @staticmethod
+    def New():
+        return teca_model_segmentation()
+
+    def __init__(self):
+        self.variable_name = "IVT"
+        self.pred_name = self.variable_name + "_PRED"
+        self.transform_fn = None
+        self.transport_fn_args = None
+        self.model = None
+        self.device = self.set_torch_device()
+
+        self.impl = teca_py.teca_programmable_algorithm.New()
+        self.impl.set_number_of_input_connections(1)
+        self.impl.set_number_of_output_ports(1)
+        self.impl.set_execute_callback(self.get_predictions_execute())
+
+    def __str__(self):
+        ms_str = 'variable_name=%s, pred_name=%d\n\n'%( \
+            self.variable_name, self.pred_name)
+
+        ms_str += 'model:\n%s\n\n'%(str(self.model))
+
+        ms_str += 'device:\n%s\n'%(str(self.device))
+
+        return ms_str
+
+    def set_variable_name(self, name):
+        """
+        set the variable name that will be inputed to the model
+        """
+        self.variable_name = name
+
+    def set_pred_name(self, name):
+        """
+        set the variable name that will be the output to the model
+        """
+        self.pred_name = name
+
+    def set_transform_fn(self, fn, *args):
+        """
+        if the data need to be transformed in a way then a function
+        could be provided to be applied on the requested data before
+        running it to the model.
+        """
+        if not hasattr(fn, '__call__'):
+            sys.stderr.write('ERROR: The provided data transform function \
```

Comment:
in Python when an error is encountered throw an exception, this will be caught in the teca Python layer and will be hooked into teca's built in error handling system.